benchbuild/utils/revision_ranges.py

```diff
@@ -0,0 +1,276 @@
+    ancestors of c_end.
+    """
+    result = [c_start]
+    result.extend(
+        git("log", "--pretty=%H", "--ancestry-path",
+            f"{c_start}..{c_end}").strip().split())
+    return result
+
+
+class AbstractRevisionRange(abc.ABC):
+    """
+    Revision ranges represent a set of revisions, i.e., a sub-graph of the
+    history of a git repository.
+    """
+    def __init__(self, comment: tp.Optional[str] = None):
+        self.__comment = comment
+
+    @property
+    def comment(self) -> tp.Optional[str]:
+        """
+        A comment associated with this revision range.
+        """
+        return self.__comment
+
+    @abc.abstractmethod
+    def init_cache(self, repo_path: Path) -> None:
+        """
+        Subclasses relying on complex functionality for determining their set
+        of revisions can use this method to initialize a cache.
+
+        Args:
+            repo_path: The path to the git repository this range is defined for.
+        """
+
+    @abc.abstractmethod
+    def __iter__(self) -> tp.Iterator[str]:
+        """
+        Allows iterating over the revisions contained in this revision range in
+        no particular order.
+
+        Returns:
+            An iterator over the revisions contained in this revision range.
+        """
+        pass
+
+
+class SingleRevision(AbstractRevisionRange):
+    """
+    A single revision.
+
+    Args:
+        rev_id: The commit hash of the single revision.
+        comment: See :func:`AbstractRevisionRange.comment()`.
+    """
+    def __init__(self, rev_id: str, comment: tp.Optional[str] = None):
+        super().__init__(comment)
+        self.__id = rev_id
+
+    def init_cache(self, repo_path: Path) -> None:
+        pass
+
+    def __iter__(self) -> tp.Iterator[str]:
+        return [self.__id].__iter__()
+
+
+class RevisionRange(AbstractRevisionRange):
+    """
+    The range of revisions between two commits `start` and `end`, i.e., all
+    revisions that are both descendants of `start`, and ancestors of `end`.
+
+    Args:
+        id_start: The commit hash of `start`.
+        id_end: The commit hash of `end`.
+        comment: See :func:`AbstractRevisionRange.comment()`.
+    """
+    def __init__(self,
+                 id_start: str,
+                 id_end: str,
+                 comment: tp.Optional[str] = None):
+        super().__init__(comment)
+        self.__id_start = id_start
+        self.__id_end = id_end
+        # cache for commit hashes
+        self.__revision_list: tp.Optional[tp.List[str]] = None
+
+    def init_cache(self, repo_path: Path) -> None:
+        git = _get_git_for_path(repo_path)
+        self.__revision_list = _get_all_revisions_between(
+            self.__id_start, self.__id_end, git)
+
+    def __iter__(self) -> tp.Iterator[str]:
+        if self.__revision_list is None:
+            raise AssertionError
+        return self.__revision_list.__iter__()
+
+
+class CommitState(IntFlag):
+    BOT = 0
```

Comment:
Are you using BOT as the bottom elem or is this something else?

Reply:
BOT means bottom here. I can rename this.

Reply:
I thought so :)